Integrate Balancing Strategy for Partitioning with Vertex Weights

N-Maas opened the pull request:
Integrate the results of my bachelor-thesis on hypergraph partitioning with vertex weights into KaHyPar. The main contribution is a balancing strategy with a theoretical guarantee for a balanced partition (given that the used bin packing algorithm is able to produce a sufficiently balanced solution), which nonetheless doesn't degrade result quality more than necessary. This strategy replaces the one that is currently used, which restarts the partitioning process several times if the result is not balanced. Additionally, a bin packing initial partitioner is added to the portfolio of initial partitioning algorithms.

In more detail, the pull request contains the following changes:
- the algorithmic core of the feature located at kahypar/partition/bin_packing, including a lot of unit tests
- a segment tree data structure for efficient range queries which is required by the algorithm (thanks to @kittobi1992 for the original implementation)
- integration of the new algorithm into the recursive bisection algorithm
- new command line and context parameters, which are also added to the recent configs
- adjustments to the fixed vertex handling of the coarsening algorithms, which doesn't work correctly at the moment and specifically violates invariants of the priority queue based coarseners
- a bin packing initial partitioning algorithm
- 2 tools for creating weighted instances from bookshelf/.mtx files (the former using the area of a component as weight, the latter using the row-net model where the vertex weight is equal to the degree of the vertex)
- some incidental smaller fixes which are not directly related to the other changes

N-Maas commented:
I tagged some locations with a TODO where I'd appreciate some feedback to the described issue.

N-Maas commented:
Note that this PR does not support individual part weights yet. I plan to open a new PR for it when the implementation is done or alternatively I could push the necessary changes to this PR, if preferred.

N-Maas commented:
Also I'm a bit unsure about the various tests. I updated a few where it was obviously necessary. However, there are tests that are failing due to outdated config files (which is also an issue on current master, I believe).

SebastianSchlag commented:
> Also I'm a bit unsure about the various tests. I updated a few where it was obviously necessary. However, there are tests that are failing due to outdated config files (which is also an issue on current master, I believe).

On master, there are no failing tests at the moment: https://travis-ci.org/github/kahypar/kahypar/builds/727611008

SebastianSchlag commented:
The `~11%` decrease in test coverage also seems to indicate that there are some serious problems in this PR.

SebastianSchlag commented:
It would be good to have at least one integration test that uses this feature (see: https://github.com/kahypar/kahypar/blob/master/tests/end_to_end/kahypar_integration_tests.cc).

SebastianSchlag commented on kahypar/application/command_line_options.h:
```diff
@@ -383,6 +383,24 @@ po::options_description createInitialPartitioningOptionsDescription(Context& con
+    "(default: worst_fit)")
+    ("i-bp-early-restart",
+    po::value<bool>(&context.initial_partitioning.infeasible_early_restart)->value_name("<bool>"),
+    "Enable early restart with prepacking of current bisection if infeasible"
```
if -> is

SebastianSchlag commented on kahypar/datastructure/hypergraph.h:
```diff
@@ -1443,6 +1443,16 @@ class GenericHypergraph {
   }
 
+  // ! Removes all fixed vertices
+  void resetFixedVertices() {
```
Can you add a test case for this method in `kahypar/tests/datastructure/hypergraph_test.cc`?

N-Maas replied:
Done

SebastianSchlag commented on kahypar/partition/coarsening/policies/fixed_vertex_acceptance_policy.h:
```diff
@@ -56,16 +56,6 @@ class FixedVertexAcceptancePolicy : public meta::PolicyBase {
     const HypernodeWeight node_weight = hg.isFixedVertex(v) ? 0 : hg.nodeWeight(v);
     return fixed_vertex_part_weight + node_weight <= max_allowed_fixed_vertex_part_weight;
   }
-
```
@kittobi1992 ?

N-Maas replied:
The coarsening algorithms are now based on the number of free vertices, therefore the additional test in FixedVertexAcceptancePolicy is not required anymore.

SebastianSchlag commented:
@N-Maas Did you discuss the fixed-vertex-related changes in the coarsening algorithms with @kittobi1992 ?

N-Maas commented on kahypar/partition/bin_packing/algorithms.h:
```diff
@@ -0,0 +1,174 @@
+#include "kahypar/datastructure/binary_heap.h"
+
+namespace kahypar {
+namespace bin_packing {
```
added to `kahypar/tests/partitionpartition/bin_packing_test.cc`

N-Maas commented on kahypar/datastructure/hypergraph.h:
```diff
@@ -1614,6 +1624,10 @@ class GenericHypergraph {
     return _fixed_vertices ? _fixed_vertices->size() : 0;
   }
 
+  HypernodeID currentNumFreeVertices() const {
```
Done

N-Maas commented on kahypar/datastructure/segment_tree.h:
```diff
@@ -0,0 +1,143 @@
+namespace kahypar {
+namespace ds {
+template<typename S, typename T, typename... Rs>
+struct segtree {
```
I'm unsure what you mean here, there are tests in `kahypar/tests/datastructure/segment_tree_test.cc`.

SebastianSchlag replied:
Okay, I may have missed those. :)

N-Maas commented:
> > Also I'm a bit unsure about the various tests. I updated a few where it was obviously necessary. However, there are tests that are failing due to outdated config files (which is also an issue on current master, I believe).
> 
> On master, there are no failing tests at the moment: https://travis-ci.org/github/kahypar/kahypar/builds/727611008

Well, I'm unsure what the CI is doing exactly, but when running locally `make integration_tests` on master I get `Illegal option: kway_fm_flow_km1`. Strictly speaking, this is indeed not a failing test but the testing process itself exits.

As far as I know, the cause is that this refinement algorithm was removed with PR #57 (Hyperflow cutter integration) but is still present in old configs.

Additionally, with the integration of the progress bar `make hypergraph_io_test` in debug mode results in:
```
[metrics.h:66:getMetric]: Assertion `mode == Mode::recursive_bisection` failed:  
Aborted
```

N-Maas commented:
> It would be good to have at least one integration test that uses this feature (see: https://github.com/kahypar/kahypar/blob/master/tests/end_to_end/kahypar_integration_tests.cc).

I added an integration test with an instance where KaHyPar currently is unable to find a balanced solution. Unfortunately, the test is a bit more expensive with k=32, but this is the smallest such real-world instance I know (ISPD98_ibm16). Alternatively, I could use an artificial instance with k=8.

N-Maas commented:
> @N-Maas Did you discuss the fixed-vertex-related changes in the coarsening algorithms with @kittobi1992 ?

Not the code changes in detail, but we discussed the overall problem and the solution I chose. Basically, the method `acceptFixedVertexImbalance` in `FixedVertexAcceptancePolicy` evaluates the validity of a contraction based on the _global_ property of the imbalance of the fixed vertex subgraph. Thus, a performed fixed vertex contraction might invalidate or revalidate arbitrary other contractions that are currently in the PQ. On the other hand, the updating strategy of the PQ-based coarseners relies on the _locality_ of the change caused by a contraction to only update adjacent hypernodes.

Therefore, there are two possible failure modes:
- A non-adjacent node is invalidated by a fixed vertex contraction and thus invalid when popped from the PQ. This case is catched by the additional check I inserted.
- A possible contraction of a non-adjacent node is revalidated, possibly being the best available for this node. Unfortunately, there seems to be no efficient way to test for this case, thus the PQ-based coarseners might miss a better contraction possibility in such a case.

N-Maas commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,166 @@
+    } else if (level == BalancingLevel::guaranteed) {
+      for (size_t i = 0; i < static_cast<size_t>(context.initial_partitioning.k); ++i) {
+        HypernodeWeight lower = context.initial_partitioning.perfect_balance_partition_weight[i];
+        HypernodeWeight& border = context.initial_partitioning.upper_allowed_partition_weight[i];
```
Yes it is, I tried to adjust the code to make it more clear.

N-Maas commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,166 @@
+        // The basic problem is, if <border> (the allowed partition weight) is too close to <upper>, the prepacking
+        // algorithm performs badly in the sense that almost every vertex will be prepacked, thus resulting in a
+        // bisection which is indeed balanced, but probably has a really poor cut.
+        if (upper - border < (border - lower) / 10) {
```
Unfortunately no, as I had no good idea how a good experimental setup for these numbers could look like and there were many other things that needed to be finished for my thesis. The numbers are just some values that seemed sensible based on my theoretical understanding of the algorithm.

The problem is that the prepacking algorithm is usually called for no more than 1 or 2 bisections on the real-world instances I used, possibly not invoking this case at all. Therefore, I suspect a really fine tuned setup is required to measure any notable effect on result quality (perhaps it might help a bit to turn k-way refinement off and only evaluate the initial partitioning result).

kittobi1992 commented:
> > > Also I'm a bit unsure about the various tests. I updated a few where it was obviously necessary. However, there are tests that are failing due to outdated config files (which is also an issue on current master, I believe).
> > 
> > 
> > On master, there are no failing tests at the moment: https://travis-ci.org/github/kahypar/kahypar/builds/727611008
> 
> Well, I'm unsure what the CI is doing exactly, but when running locally `make integration_tests` on master I get `Illegal option: kway_fm_flow_km1`. Strictly speaking, this is indeed not a failing test but the testing process itself exits.
> 
> As far as I know, the cause is that this refinement algorithm was removed with PR #57 (Hyperflow cutter integration) but is still present in old configs.
> 
> Additionally, with the integration of the progress bar `make hypergraph_io_test` in debug mode results in:
> 
> ```
> [metrics.h:66:getMetric]: Assertion `mode == Mode::recursive_bisection` failed:  
> Aborted
> ```

Yes, this was my fault. I thought that I am able to integrate some easy features without breaking the test suite => I was wrong :-D

kittobi1992 commented:
> > @N-Maas Did you discuss the fixed-vertex-related changes in the coarsening algorithms with @kittobi1992 ?
> 
> Not the code changes in detail, but we discussed the overall problem and the solution I chose. Basically, the method `acceptFixedVertexImbalance` in `FixedVertexAcceptancePolicy` evaluates the validity of a contraction based on the _global_ property of the imbalance of the fixed vertex subgraph. Thus, a performed fixed vertex contraction might invalidate or revalidate arbitrary other contractions that are currently in the PQ. On the other hand, the updating strategy of the PQ-based coarseners relies on the _locality_ of the change caused by a contraction to only update adjacent hypernodes.
> 
> Therefore, there are two possible failure modes:
> 
> * A non-adjacent node is invalidated by a fixed vertex contraction and thus invalid when popped from the PQ. This case is catched by the additional check I inserted.
> * A possible contraction of a non-adjacent node is revalidated, possibly being the best available for this node. Unfortunately, there seems to be no efficient way to test for this case, thus the PQ-based coarseners might miss a better contraction possibility in such a case.

I already had a discussion about that with @SebastianSchlag. We should fix the first case and ignore the second. Further, we should validate if we can replace the LazyVertexCoarsener with the MLCoarsener in RB-Mode, because such a experiment was never executed.

kittobi1992 commented on kahypar/partition/bin_packing/bin_packing.h:
```diff
@@ -0,0 +1,458 @@
+  };
+
+  // Returns the current hypernodes sorted in descending order of weight.
+  static inline std::vector<HypernodeID> extractNodesWithDescendingWeight(const Hypergraph& hg) {
```
consider adding a parameter that indicates wheater nodes with same weight should be random shuffled or not

kittobi1992 commented on kahypar/partition/bin_packing/bin_packing.h:
```diff
@@ -0,0 +1,458 @@
+  };
+
+  // Returns the current hypernodes sorted in descending order of weight.
+  static inline std::vector<HypernodeID> extractNodesWithDescendingWeight(const Hypergraph& hg) {
```
sortVerticesInDescendingNodeWeightOrder?

N-Maas replied:
renamed to nodesInDescendingWeightOrder (sortVerticesInDescendingNodeWeightOrder sounds like the hypergraph is mutated)

kittobi1992 commented on kahypar/partition/bin_packing/bin_packing.h:
```diff
@@ -0,0 +1,458 @@
+  }
+
+
+  // segment tree that is used to ensure the balance of a prepacking
```
balance_max, balance_base, BalanceSegTree and getMaxPartIndex should be in a anonymous namespace
namespace {
 ....
}

N-Maas commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,167 @@
+    return previous;
+  }
+
+class IBinPacker {
```
as discussed, it is necessary

N-Maas commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,167 @@
+    return previous;
+  }
+
+class IBinPacker {
```
Unfortunately, it is not only responsible for prepacking, but also used by the bin packing initial partitioner (i.e. for all algorithms where a bin packing algorithm is selected). For individual part weights, even more methods will be added.

N-Maas commented on kahypar/partition/recursive_bisection.h:
```diff
@@ -52,11 +53,15 @@ class RBState {
     hypergraph(std::move(h)),
     state(s),
+    level(BalancingLevel::none),
+    isFeasible(true),
```
is_feasible (the value indicates whether a balanced final partition is possible, not whether the current partition is balanced)

N-Maas commented on kahypar/partition/recursive_bisection.h:
```diff
@@ -211,16 +240,42 @@ static inline void partition(Hypergraph& input_hypergraph,
+
+            bool balanced = true;
+            for (PartitionID i = k1; i <= k2; ++i) {
+              if (input_hypergraph.partWeight(i) > lmax) {
```
yeah, this will be adressed in the individual part weights PR

N-Maas commented on kahypar/partition/multilevel.h:
```diff
@@ -124,5 +128,45 @@ static inline void partition(Hypergraph& hypergraph,
   io::printLocalSearchResults(context, hypergraph);
 }
+
+static inline void partitionRepeatedOnInfeasible(Hypergraph& hypergraph,
```
Hm, it depends on the input parameters whether there are balance guarantees..

SebastianSchlag replied:
Can we come up with a better name for this method? :)

N-Maas replied:
Hmm I've no really good idea at the moment, maybe `partitionWithPrepacking` is a bit better?

N-Maas commented on kahypar/partition/bin_packing/prepacking.h:
```diff
@@ -0,0 +1,215 @@
+    hg.setFixedVertex(nodes[i], parts[i]);
+  }
+
+  // TODO(maas) It is questionable whether this last part of the algorithm is required.
+  // It is an optimization which theoretically allows to relax the upper allowed partition
+  // weights for the bisection in some cases. However, we also do know that more relaxed weights
+  // in a higher bisection level might lead to worse results in lower levels or in the refinement.
+  // Thus, in practice it is probably no improvement.
+  //
+  // The ugly aspect here is that it also complicates code at other positions, because the changed
+  // weights need to be propagated to the initial partitioning algorithm (see initial_partition::partition).
```
Any thoughts on this?

SebastianSchlag replied:
Could we do some experiments to check whether or not it is beneficial to use this?

N-Maas replied:
I have started an experiment comparing two variants where this is enabled/disabled. Hopefully, the results will show whether there is an impact on the result quality.

N-Maas replied:
The result showed a slightly negative impact on result quality, therefore I removed this part of the algorithm.

SebastianSchlag commented:
> > > Also I'm a bit unsure about the various tests. I updated a few where it was obviously necessary. However, there are tests that are failing due to outdated config files (which is also an issue on current master, I believe).
> > 
> > 
> > On master, there are no failing tests at the moment: https://travis-ci.org/github/kahypar/kahypar/builds/727611008
> 
> Well, I'm unsure what the CI is doing exactly, but when running locally `make integration_tests` on master I get `Illegal option: kway_fm_flow_km1`. Strictly speaking, this is indeed not a failing test but the testing process itself exits.
> 
> As far as I know, the cause is that this refinement algorithm was removed with PR #57 (Hyperflow cutter integration) but is still present in old configs.
> 
> Additionally, with the integration of the progress bar `make hypergraph_io_test` in debug mode results in:
> 
> ```
> [metrics.h:66:getMetric]: Assertion `mode == Mode::recursive_bisection` failed:  
> Aborted
> ```
@kittobi1992 Can you please take a look at the issue with the progress bar?

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,167 @@
+  BinPacker() = default;
+
+ private:
+  void prepackingImpl(Hypergraph& hg, Context& context, const BalancingLevel level) override {
```
> hypergraph and context is passed to each interface function, consider storing them as members and passing them in the constructor

As a side note to this: Please avoid passing non-const references to a context around, since this makes it harder to track config changes. If you need to change parameters, create a copy of the current context and change the corresponding parameters there.

SebastianSchlag commented:
It seems that there are still some failing test-cases because of the handling of individual part weights.

SebastianSchlag commented:
A general remark on the config changes: In older configurations, bin-packing should never be enabled by default in order to maintain compatibility with the corresponding papers.

SebastianSchlag commented on tools/hypergraph_statistics.cc:
```diff
@@ -49,7 +49,7 @@ int main(int argc, char* argv[]) {
   kahypar::io::readHypergraphFile(graph_filename, num_hypernodes, num_hyperedges,
                                   index_vector, edge_vector, &hyperedge_weights, &hypernode_weights);
-  Hypergraph hypergraph(num_hypernodes, num_hyperedges, index_vector, edge_vector);
+  Hypergraph hypergraph(num_hypernodes, num_hyperedges, index_vector, edge_vector, 2, hyperedge_weights, hypernode_weights);
```
Just for the sake of completeness, could you please also extend this to include the same statistics for hyperedge weights?

N-Maas replied:
Done

SebastianSchlag commented on config/old_reference_configs/cut_kKaHyPar_dissertation.ini:
```diff
@@ -45,6 +45,10 @@ i-c-fixed-vertex-acceptance-criterion=fixed_vertex_allowed
 # initial partitioning -> initial partitioning
 i-algo=pool
 i-runs=20
+# initial partitioning -> bin packing
```
For example in this config, we shouldn't do bin packing. However, I see that we might run into problems here, because the old code that @kittobi1992 wrote that triggers the repartitioning with tighter balance constraint, is no longer available after merging this. We therefore should create a backup branch that contains the current code, before we merge this PR.

N-Maas replied:
I removed the parameters from the old configs and instead adjusted the test cases manually. A backup branch seems reasonable, but currently I dont have the write permissions to create one.

N-Maas commented:
> It seems that there are still some failing test-cases because of the handling of individual part weights.

Should be fixed now by disabling the bin packing features for these tests. Using bin packing in combination with individual part weights will be addressed in a separate PR.

SebastianSchlag commented on kahypar/datastructure/segment_tree.h:
```diff
@@ -0,0 +1,97 @@
+         T (*BASE)(const size_t &, const std::vector<S> &, const T &)>
+class ParametrizedSegmentTree {
+public:
+  typedef S seq_type;
```
Just as a general remark: prefer `using` over typedefs: `using seq_type = S`

SebastianSchlag commented on kahypar/datastructure/segment_tree.h:
```diff
@@ -0,0 +1,97 @@
+  typedef S seq_type;
+  typedef T tree_type;
+
+  explicit ParametrizedSegmentTree(std::vector<seq_type>& seq, T param) : _size(seq.size()), _seq(seq), _seg_tree(2*_size), _param(param) {
```
Consider using `tree_type` instead of `T` here.

SebastianSchlag commented on kahypar/datastructure/segment_tree.h:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  size_t _size;
+  std::reference_wrapper<std::vector<S>> _seq;
```
Consider using `tree_type` instead of `T`  and `seq_type` instead of `S` here and below.

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,124 @@
+    const PartitionID rb_range_k = _context.partition.rb_upper_k - _context.partition.rb_lower_k + 1;
+    const HypernodeWeight max_bin_weight = floor(_context.initial_partitioning.current_max_bin_weight * (1.0 + _context.initial_partitioning.bin_epsilon));
+
+    if (level == BalancingLevel::heuristic) {
```
By using `kahypar/kahypar/meta/int_to_type.h` you could get rid of that if and have an overload of `prepackingImpl` for each BalancingLevel type.

N-Maas replied:
I don't understand how this works, wouldn't I still need to use either dynamic dispatch or a case distinction to select the appropriate overload?

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,124 @@
+    const HypernodeWeight max_bin_weight = floor(_context.initial_partitioning.current_max_bin_weight * (1.0 + _context.initial_partitioning.bin_epsilon));
+
+    if (level == BalancingLevel::heuristic) {
+      bin_packing::calculateHeuristicPrepacking<BPAlgorithm>(_hypergraph, _context, rb_range_k, max_bin_weight);
```
`bin_packing::` shouldn't be necessary here, since you are already in the `bin_packing` namespace.

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,124 @@
+        }
+      }
+
+      bin_packing::calculateExactPrepacking<BPAlgorithm>(_hypergraph, _context, rb_range_k, max_bin_weight);
```
`bin_packing::` shouldn't be necessary here, since you are already in the `bin_packing` namespace.

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,124 @@
+    TwoLevelPacker<BPAlgorithm> packer(rb_range_k, max_bin_weight);
+
+    if (_hypergraph.containsFixedVertices()) {
+      bin_packing::preassignFixedVertices<BPAlgorithm>(_hypergraph, nodes, parts, packer, _context.partition.k, rb_range_k);
```
`bin_packing::` shouldn't be necessary here, since you are already in the `bin_packing` namespace.

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packer.h:
```diff
@@ -0,0 +1,124 @@
+  Context& _context;
+};
+
+static std::unique_ptr<IBinPacker> createBinPacker(const BinPackingAlgorithm& bp_algo, Hypergraph& hypergraph, Context& context) {
```
For consistency, I would propose to use our `Factory` approach here (see: `kahypar/partition/factories.h`). @kittobi1992 can help you with that, if you run into any problems.

N-Maas commented:
Sorry that I didn't continue working on the PR as the focus shifted to the recent paper.

I have now included the use of factories for the selection of the bin packing algorithm. Further, I added some small changes to the prepacking algorithm itself which were introduced during the work on the paper (specifically, the formulation of the balance constraint) and updated the RemvoveHeavyNodes utility to the approach used in the paper.

N-Maas commented on kahypar/partition/initial_partition.h:
```diff
@@ -155,64 +137,53 @@ static inline Context createContext(const Hypergraph& hg,
+    // NOTE: If the fixed vertex subgraph is imbalanced, we cannot guarantee
+    //       a balanced initial partition with recursive bisection. Therefore,
+    //       we switch to direct k-way flat mode in that case.
+    if (!(context.initial_partitioning.technique == InitialPartitioningTechnique::flat &&
```
I have adjusted the check here such that the warning ("Fixed vertex subgraph is imbalanced") may only be triggered before initial partitioning. Previously, the warning sometimes was (rather inappropriately) triggered by the fixed vertices of the prepacking algorithm.

However, I guess we could the remove the `fixed_vertex_subgraph_imbalance > context.partition.epsilon` check, the warning and the switching to flat mode altogether, as with the inclusion of this PR there is no more reason to believe that flat partitioning provides better balance than recursive bisection.

SebastianSchlag replied:
Should we do some small experiments to verify that or can we already say that based on the experiments we did so far?

N-Maas replied:
About the balance guarantees of the recursive bisection algorithm I don't have much doubt. After all they are the main point of the upcoming paper, including a formal proof. I.e. the RB algorithm can find a balanced partition if the used load balancing algorithm can find one. Flat partitioning can hardly do better, except of course implementing a better load balancing algorithm.

SebastianSchlag commented:
> > It would be good to have at least one integration test that uses this feature (see: https://github.com/kahypar/kahypar/blob/master/tests/end_to_end/kahypar_integration_tests.cc).
> 
> I added an integration test with an instance where KaHyPar currently is unable to find a balanced solution. Unfortunately, the test is a bit more expensive with k=32, but this is the smallest such real-world instance I know (ISPD98_ibm16). Alternatively, I could use an artificial instance with k=8.

Using an artificial instance is probably better if it decreases the running time.

SebastianSchlag commented:
> Sorry that I didn't continue working on the PR as the focus shifted to the recent paper.
> 
> I have now included the use of factories for the selection of the bin packing algorithm. Further, I added some small changes to the prepacking algorithm itself which were introduced during the work on the paper (specifically, the formulation of the balance constraint) and updated the RemvoveHeavyNodes utility to the approach used in the paper.

No worries. Working on the paper was clearly the correct priority. I'm going to do a final review of the PR before merging. Also, I already created a bugfix release for the current state (https://github.com/kahypar/kahypar/releases/tag/1.2.1). This will then be the last release that did not contain your changes.

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packing_algorithms.h:
```diff
@@ -0,0 +1,171 @@
+};
+
+// First Fit algorithm - inserts an element to the first fitting bin.
+class FirstFit {
```
Do we still need both implementations or do we only use one? If so, it might make sense to remove the unused implementation to reduce unused code.

N-Maas replied:
Well, it depends. Worst fit is the default, but in my thesis there were also a few cases where first fit performed better.

(It can be selected with the i-bp-algorithm parameter.)

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packing_algorithms.h:
```diff
@@ -0,0 +1,171 @@
+
+  private:
+    HypernodeWeight _max_bin_weight;
+    std::vector<std::pair<HypernodeWeight, bool>> _bins;
```
`#include <utility>`

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packing_utils.h:
```diff
@@ -0,0 +1,270 @@
+  }
+
+ private:
+  std::vector<PartitionID> _parts;
```
`#include <vector>`

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packing_utils.h:
```diff
@@ -0,0 +1,270 @@
+
+  // Packs the current bins into the specified parts. Returns the partition mapping for the bins
+  // and a vector of the resulting part weights.
+  std::pair<PartitionMapping, std::vector<HypernodeWeight>> applySecondLevel(const std::vector<HypernodeWeight>& max_allowed_part_weights,
```
`#include <utility>`

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packing_utils.h:
```diff
@@ -0,0 +1,270 @@
+    std::vector<PartitionID> bin_counts(max_allowed_part_weights.size(), 0);
+    PartitionMapping mapping = _bins_to_parts;
+
+    const HypernodeWeight max_part_weight = *std::max_element(max_allowed_part_weights.cbegin(),
```
`#include <algorithm>`

SebastianSchlag commented on kahypar/partition/bin_packing/bin_packing_utils.h:
```diff
@@ -0,0 +1,270 @@
+
+    // sort bins in descending order and calculate the packing
+    std::vector<PartitionID> kbins_descending(_alg.numBins());
+    std::iota(kbins_descending.begin(), kbins_descending.end(), 0);
```
`#include <numeric>`

SebastianSchlag commented on kahypar/partition/bin_packing/i_bin_packer.h:
```diff
@@ -0,0 +1,82 @@
+namespace kahypar {
+namespace bin_packing {
+// Represents the level of the applied prepacking.
+enum class BalancingLevel : uint8_t {
```
Should this be stored in `kahypar/kahypar/partition/context_enum_classes.h`?

SebastianSchlag replied:
Maybe also add a brief description for each of the enum values.

N-Maas replied:
I'm not sure. Unlike the enums in `kahypar/kahypar/partition/context_enum_classes.h` it is neither stored in the context (which would not work because the context of a bisection is not saved) nor is a conversion to or from a string required.

SebastianSchlag replied:
I see, good points!

SebastianSchlag commented on kahypar/partition/bin_packing/i_bin_packer.h:
```diff
@@ -0,0 +1,82 @@
+
+ private:
+  virtual void prepackingImpl(const BalancingLevel level) = 0;
+  virtual std::vector<PartitionID> twoLevelPackingImpl(const std::vector<HypernodeID>& nodes, const HypernodeWeight max_bin_weight) const = 0;
```
`#include <vector>`

SebastianSchlag commented on kahypar/partition/bin_packing/i_bin_packer.h:
```diff
@@ -0,0 +1,82 @@
+namespace kahypar {
+namespace bin_packing {
+// Represents the level of the applied prepacking.
+enum class BalancingLevel : uint8_t {
```
`#include <cstdint>`

SebastianSchlag commented on kahypar/partition/bin_packing/prepacking.h:
```diff
@@ -0,0 +1,186 @@
+    return std::max(v1, v2);
+  }
+
+  HypernodeWeight balance_base(const size_t& i, const std::vector<std::pair<HypernodeWeight, HypernodeWeight>>& seq, const HypernodeWeight& k) {
```
`#include <vector>`
`#include <utility>`

SebastianSchlag commented on kahypar/partition/bin_packing/prepacking.h:
```diff
@@ -0,0 +1,186 @@
+namespace kahypar {
+namespace bin_packing {
+namespace {
+  // segment tree that is used to ensure the balance of a prepacking
```
That comment seems to be wrong.

N-Maas replied:
Yeah, at least misleading

SebastianSchlag commented on kahypar/partition/bin_packing/prepacking.h:
```diff
@@ -0,0 +1,186 @@
+  ASSERT(part_weight.size() == upper_weight.size() && part_weight.size() == num_bins_per_part.size());
+
+  size_t max_imb_part_id = 0;
+  HypernodeWeight min_remaining = std::numeric_limits<HypernodeWeight>::max();
```
`#include <limits>`

SebastianSchlag commented on kahypar/partition/bin_packing/prepacking.h:
```diff
@@ -0,0 +1,186 @@
+    const HypernodeWeight searched_weight = weights[i].second + remaining_weight_for_max_part;
+
+    // find subrange of specified weight
+    const size_t j = std::lower_bound(weights.cbegin() + i, weights.cend(), std::make_pair(0, searched_weight),
```
`#include <algorithm>`

SebastianSchlag commented on kahypar/partition/bin_packing/prepacking.h:
```diff
@@ -0,0 +1,186 @@
+}
+
+template< class BPAlgorithm >
+static inline void calculateHeuristicPrepacking(Hypergraph& hg, const Context& context, const PartitionID& rb_range_k, const HypernodeWeight max_bin_weight) {
```
Do we still use both `calculateHeuristicPrepacking` and `calculateHeuristicPrepacking`?

N-Maas replied:
The heuristic prepacking is not used in the paper, but I think it might make sense to keep it. The reason we do not use it in the paper is to keep things simpler in the theoretical description. Also we ran some preliminary experiments, with the result that using the heuristic prepacking improves the result quality, but the difference was quite small so it is no problem to exclude it in the paper.

I have updated the code now to include an additional parameter `i-bp-heuristic-prepacking` which enables/disables the heuristic prepacking (which is the same I did for the experiments). For the current configs I disabled it to match the version of the paper.

SebastianSchlag commented on kahypar/partition/coarsening/vertex_pair_coarsener_base.h:
```diff
@@ -81,8 +81,8 @@ class VertexPairCoarsenerBase : public CoarsenerBase {
     if (_context.type == ContextType::main) {
-      _context.stats.set(StatTag::InitialPartitioning, "inititalCut", current_metrics.cut);
-      _context.stats.set(StatTag::InitialPartitioning, "inititalKm1", current_metrics.km1);
+      _context.stats.set(StatTag::InitialPartitioning, "initialCut", current_metrics.cut);
```
👍 👍 👍 👍

SebastianSchlag commented on kahypar/partition/context.h:
```diff
@@ -657,6 +683,11 @@ static inline void sanityCheck(const Hypergraph& hypergraph, Context& context) {
       LOG << "Sum of individual part weights is less than sum of vertex weights";
       std::exit(-1);
     }
+    if (context.initial_partitioning.enable_early_restart || context.initial_partitioning.enable_late_restart) {
```
How complicated would it be to get support for individual part weights?

kittobi1992 replied:
he is currently working on that and it will be integrated in separate PR

SebastianSchlag replied:
Awesome! That's good news! 👍

SebastianSchlag commented on kahypar/partition/context_enum_classes.h:
```diff
@@ -112,10 +112,17 @@ enum class InitialPartitionerAlgorithm : uint8_t {
 
+enum class BinPackingAlgorithm : uint8_t {
+  worst_fit,
+  first_fit,
```
As above: If only one of those is used, we should get rid of the unused one.

SebastianSchlag commented on kahypar/partition/factories.h:
```diff
@@ -52,6 +53,9 @@ using InitialPartitioningFactory = meta::Factory<InitialPartitionerAlgorithm,
 using RefinerFactory = meta::Factory<RefinementAlgorithm,
                                      IRefiner* (*)(Hypergraph&, const Context&)>;
 
+using BinPackerFactory = meta::Factory<BinPackingAlgorithm,
```
👍

SebastianSchlag commented on kahypar/partition/initial_partitioning/bin_packing_initial_partitioner.h:
```diff
@@ -0,0 +1,141 @@
+
+    const PartitionID rb_range_k = _context.partition.rb_upper_k - _context.partition.rb_lower_k + 1;
+    const HypernodeWeight allowed_part_weight = (1.0 + _context.partition.epsilon) * (static_cast<double>(_hg.totalWeight()) / rb_range_k);
+    const std::vector<PartitionID> parts = _bin_packer->twoLevelPacking(_descending_nodes, allowed_part_weight);
```
`#include <vector>`

SebastianSchlag commented on kahypar/partition/initial_partitioning/bin_packing_initial_partitioner.h:
```diff
@@ -0,0 +1,141 @@
+  using Base::_context;
+
+  std::vector<HypernodeID> _descending_nodes;
+  std::unique_ptr<IBinPacker> _bin_packer;
```
`#include <memory>`

SebastianSchlag commented on kahypar/partition/initial_partitioning/greedy_hypergraph_growing_initial_partitioner.h:
```diff
@@ -117,7 +117,7 @@ class GreedyHypergraphGrowingInitialPartitioner : public IInitialPartitioner,
       _pq.disablePart(_context.initial_partitioning.unassigned_part);
       minimum_unassigned_part_weight =
-        _context.initial_partitioning.perfect_balance_partition_weight[unassigned_part];
+        _context.initial_partitioning.perfect_balance_partition_weight[_context.initial_partitioning.unassigned_part];
```
Was that a bug?

N-Maas replied:
Yes, I remember that there was a case where I got an out of bounds access (unassigned_part = -1).

SebastianSchlag commented on kahypar/partition/initial_partitioning/initial_partitioner_base.h:
```diff
@@ -86,8 +86,6 @@ class InitialPartitionerBase {
     }
     _context.partition.max_part_weights =
       _context.initial_partitioning.upper_allowed_partition_weight;
-    _context.partition.max_part_weights =
```
👍

SebastianSchlag commented on kahypar/partition/metrics.h:
```diff
@@ -194,7 +194,7 @@ static inline double avgHypernodeWeight(const Hypergraph& hypergraph) {
     sum += hypergraph.nodeWeight(hn);
   }
-  return sum / hypergraph.currentNumNodes();
+  return static_cast<double>(sum) / hypergraph.currentNumNodes();
```
👍

SebastianSchlag commented on kahypar/partition/multilevel.h:
```diff
@@ -124,5 +126,45 @@ static inline void partition(Hypergraph& hypergraph,
+
+static inline void partitionRepeatedOnInfeasible(Hypergraph& hypergraph,
+                                                 const Context& context,
+                                                 Context::PartitioningStats& stats,
```
Is there a specific reason that the stats are passed as an additional non-const reference? The stats of the context object are mutable so they can be changed even if the context is passed in as const ref.

N-Maas replied:
Yes. To this method we pass not the original context but the context of the current bisection. I can use the stats of the current context, but then the information is lost after the current bisection. Thus the stats of the original context need to be passed additionally.

Of course a possible solution could be to share the stats between copies of the context (via shared_ptr?), but this would probably require changes at other locations, too.

SebastianSchlag replied:
Okay, I see. In that case, your approach totally makes sense!

SebastianSchlag commented on kahypar/partition/multilevel.h:
```diff
@@ -124,5 +126,45 @@ static inline void partition(Hypergraph& hypergraph,
+    }
+
+    // perform prepacking of heavy vertices
+    std::unique_ptr<IBinPacker> bin_packer(
```
👍

SebastianSchlag replied:
`#include <memory>`

SebastianSchlag commented on kahypar/partition/recursive_bisection.h:
```diff
@@ -211,16 +240,42 @@ static inline void partition(Hypergraph& input_hypergraph,
+            }
+          }
+
+          hypergraph_stack.pop_back();
```
What about this comment from @kittobi1992 ?

N-Maas replied:
Good point, is addressed now

SebastianSchlag commented on kahypar/partition/registries/register_bin_packing_algorithms.h:
```diff
@@ -0,0 +1,45 @@
+#include "kahypar/partition/context.h"
+#include "kahypar/partition/factories.h"
+
+#define REGISTER_BIN_PACKER(id, bin_packer)                               \
```
👍

SebastianSchlag commented on kahypar/utils/stats.h:
```diff
@@ -84,7 +84,7 @@ class Stats {
 
   void add(const StatTag& tag, const std::string& key, const double& value) {
-    _logs[static_cast<size_t>(tag)][key] = value;
+    _logs[static_cast<size_t>(tag)][key] += value;
```
🤡

kittobi1992 replied:
:-D

SebastianSchlag commented on tests/datastructure/segment_tree_test.cc:
```diff
@@ -0,0 +1,158 @@
+
+// Naive Range Minimum Query
+int naive_min(const size_t& i, const size_t& j, const std::vector<int>& seq, const int& /*param*/) {
+  int min = std::numeric_limits<int>::max();
```
`#include <limits>`

SebastianSchlag commented on tests/datastructure/segment_tree_test.cc:
```diff
@@ -0,0 +1,158 @@
+}
+
+// Naive Range Minimum Query
+int naive_min(const size_t& i, const size_t& j, const std::vector<int>& seq, const int& /*param*/) {
```
`#include <vector>`

SebastianSchlag commented on tests/datastructure/segment_tree_test.cc:
```diff
@@ -0,0 +1,158 @@
+    ASequence() :
+    sequence() {
+      std::mt19937 gen(std::random_device{}());
+      int seed = std::uniform_int_distribution<>{ 0 }(gen);
```
`#include <random>`

SebastianSchlag commented on tests/datastructure/segment_tree_test.cc:
```diff
@@ -0,0 +1,158 @@
+int naive_sum_of_powers(const size_t& i, const size_t& j, const std::vector<int>& seq, const int& exp) {
+  int sum = 0;
+  for (size_t t = i; t <= j; ++t) {
+    sum += std::pow(seq[t], exp);
```
`#include <cmath>`

SebastianSchlag commented on tests/end_to_end/configs/test.ini:
```diff
@@ -52,7 +52,7 @@ i-r-runs=-1
 i-r-fm-stop-i=50
 # main -> local search
-r-type=kway_fm_flow_km1
+r-type=kway_fm_km1
```
This should be `kway_fm_hyperflow_cutter`

N-Maas replied:
updated

SebastianSchlag commented on tests/partition/bin_packing_test.cc:
```diff
@@ -0,0 +1,693 @@
+    }
+
+    void initialize(const HypernodeWeightVector& weights,
+                    const std::vector<PartitionID>& partitions,
```
`#include <vector>`

SebastianSchlag commented:
One thing we should do if the paper gets accepted is to rename the configs from `config/*_sea20.ini` to `config/*_sea21.ini`

N-Maas commented:
addressed missing includes

N-Maas commented:
> > > It would be good to have at least one integration test that uses this feature (see: https://github.com/kahypar/kahypar/blob/master/tests/end_to_end/kahypar_integration_tests.cc).
> > 
> > 
> > I added an integration test with an instance where KaHyPar currently is unable to find a balanced solution. Unfortunately, the test is a bit more expensive with k=32, but this is the smallest such real-world instance I know (ISPD98_ibm16). Alternatively, I could use an artificial instance with k=8.
> 
> Using an artificial instance is probably better if it decreases the running time.

Ok, I'm using a version of `ISPD98_ibm01` with artificial weights and k=8 now.

N-Maas commented on tests/datastructure/segment_tree_test.cc:
```diff
@@ -0,0 +1,168 @@
+
+TEST_F(ASequence, EmptyCase) {
+  sequence.resize(0);
+  RangeMinimum tree(sequence, 0);
```
Sometimes testing the simple cases helps 😅

N-Maas replied:
whoops, the review was unintentionally

N-Maas commented:
Ok, I think I have addressed the open points except for the discussions where I'm unsure about the proposed changes
